open-codegen/opengen/templates/icasadi/interface.c

```diff
@@ -309,3 +506,121 @@ int mapping_f2_function_{{ meta.optimizer_name }}(const casadi_real** arg, casad
+        allocated_r_workspace_w_cost,
+        (void*) 0);
+}
+
+
+/**
+ * Interface to auto-generated CasADi function for w1(u, p), which computes an
+ * n1-dimensional vector of scaling parameters
+ */
+int preconditioning_w1_function_{{ meta.optimizer_name }}(const casadi_real** arg, casadi_real** res) {
+    /* Array of pointers to where (u, p) are stored */
+    const casadi_real* args__[W1_SZ_ARG_{{ meta.optimizer_name | upper}}] =
+            {uxip_space,  /* :u   */
+             uxip_space + IDX_P_{{ meta.optimizer_name | upper}}};  /* :p   */
+    /* Copy given data to variable `uxip_space` */
+    copy_args_into_up_space(arg);
+    /*
+     * The result should be written in result_space_w1
+     * (memory has been allocated - see beginning of this file)
+     */
+    result_space_w1[0] = res[0];
+    /*
+     * Call auto-generated function {{meta.w_f1_function_name}}
+     * Implemented in: icasadi/extern/auto_preconditioning_functions.c
+     */
+    return {{meta.w_f1_function_name}}(
+        args__,
+        result_space_w1,
+        allocated_i_workspace_w1,
+        allocated_r_workspace_w1,
+        (void*) 0);
+}
+
+
+
+/**
+ * Interface to auto-generated CasADi function for w2(u, p), which computes an
+ * n2-dimensional vector of scaling parameters
+ */
+int preconditioning_w2_function_{{ meta.optimizer_name }}(const casadi_real** arg, casadi_real** res) {
+    /* Array of pointers to where (u, p) are stored */
+    const casadi_real* args__[W2_SZ_ARG_{{ meta.optimizer_name | upper}}] =
+            {uxip_space,  /* :u   */
+             uxip_space + IDX_P_{{ meta.optimizer_name | upper}}};  /* :p   */
+    /* Copy given data to variable `uxip_space` */
+    copy_args_into_up_space(arg);
+    /*
+     * The result should be written in result_space_w2
+     * (memory has been allocated - see beginning of this file)
+     */
+    result_space_w2[0] = res[0];
+    /*
+     * Call auto-generated function {{meta.w_f2_function_name}}
+     * Implemented in: icasadi/extern/auto_preconditioning_functions.c
+     */
+    return {{meta.w_f2_function_name}}(
+        args__,
+        result_space_w2,
+        allocated_i_workspace_w2,
+        allocated_r_workspace_w2,
+        (void*) 0);
+}
+
+
+
+/**
+ * Interface to auto-generated CasADi function for rho_1(u, theta), which computes the initial
+ * penalty parameter. Note that this is a function of u and theta = (p, w_cost, w1, w2) and the
+ * caller needs to provide p, w_cost, w1 and w2
+ */
+int preconditioning_init_penalty_function_{{ meta.optimizer_name }}(const casadi_real** arg, casadi_real** res) {
+    /* Array of pointers to where (u, p) are stored */
+    const casadi_real* args__[INIT_PENALTY_SZ_ARG_{{ meta.optimizer_name | upper}}] =
+            {uxip_space,  /* :u   */
+             uxip_space + IDX_P_{{ meta.optimizer_name | upper}}};  /* :p   */
+    /* Copy given data to variable `uxip_space` */
+    copy_args_into_upw_space(arg);
```

Comment:
I don't understand this call: does the caller have to provide *two* arguments, `u, theta` or *five* arguments `u, p, w_cost, w1, w2`?

Reply:
Looking at it again, I see that there is a mismatch between the arguments of copy_ function and the actual arguments of the calling function. Either of them has to be changed.

Currently, the initial_penalty_function defined in optimizer_builder.py is designed to have 5 arguments.

Reply:
Sorry, the  initial_penalty_function (eg: `open_mapping_initial_penalty_rosenbrock`) is designed to have 2 arguments (u, theta). Hence, the `copy_args_into_upw_space` has to be modified as it is currently implemented to have 5 arguments.

Reply:
It's not a problem to have 5 arguments